File: BirthdayBot/UserInterface/CommandsCommon.vb
Imports System.Text.RegularExpressions
Imports Discord.WebSocket
Imports NodaTime

''' <summary>
''' Common base class for common constants and variables.
''' </summary>
Friend MustInherit Class CommandsCommon
    Public Const CommandPrefix = "bb."
    Public Const GenericError = ":x: Invalid usage. Consult the help command."
    Public Const BadUserError = ":x: Unable to find user. Specify their `@` mention or their ID."
    Public Const ExpectedNoParametersError = ":x: This command does not take parameters. Did you mean to use another?"

    Delegate Function CommandHandler(param As String(), reqChannel As SocketTextChannel, reqUser As SocketGuildUser) As Task

    Protected Shared ReadOnly Property TzNameMap As Dictionary(Of String, String)
        Get
            If _tzNameMap Is Nothing Then
                ' Because IDateTimeZoneProvider.GetZoneOrNull is not case sensitive:
                ' Getting every existing zone name and mapping it onto a dictionary. Now a case-insensitive
                ' search can be made with the accepted value retrieved as a result.
                _tzNameMap = New Dictionary(Of String, String)(StringComparer.InvariantCultureIgnoreCase)
                For Each name In DateTimeZoneProviders.Tzdb.Ids
                    _tzNameMap.Add(name, name)
                Next
            End If
            Return _tzNameMap
        End Get
    End Property
    Protected Shared ReadOnly ChannelMention As New Regex("<#(\d+)>")
    Protected Shared ReadOnly UserMention As New Regex("<@\!?(\d+)>")
    Private Shared _tzNameMap As Dictionary(Of String, String) ' Value set by getter property on first read

    Protected ReadOnly Instance As BirthdayBot
    Protected ReadOnly BotConfig As Configuration
    Protected ReadOnly Discord As DiscordSocketClient

    Sub New(inst As BirthdayBot, db As Configuration)
        Instance = inst
        BotConfig = db
        Discord = inst.DiscordClient
    End Sub

    ''' <summary>
    ''' On command dispatcher initialization, it will retrieve all available commands through here.
    ''' </summary>
    Public MustOverride ReadOnly Property Commands As IEnumerable(Of (String, CommandHandler))

    ''' <summary>
    ''' Checks given time zone input. Returns a valid string for use with NodaTime.
    ''' </summary>
    ''' <param name="tzinput"></param>
    ''' <returns></returns>
    Protected Function ParseTimeZone(tzinput As String) As String
        Dim tz As String = Nothing
        If tzinput IsNot Nothing Then
            ' Just check if the input exists in the map. Get the "true" value, or reject it altogether.
            If Not TzNameMap.TryGetValue(tzinput, tz) Then
                Throw New FormatException(":x: Unknown or invalid time zone name.")
            End If
        End If
        Return tz
    End Function

    ''' <summary>
    ''' Given user input where a user-like parameter is expected, attempts to resolve to an ID value.
    ''' Input must be a mention or explicit ID. No name resolution is done here.
    ''' </summary>
    Protected Function TryGetUserId(input As String, ByRef result As ULong) As Boolean
        Dim doParse As String
        Dim m = UserMention.Match(input)
        If m.Success Then
            doParse = m.Groups(1).Value
        Else
            doParse = input
        End If

        Dim resultVal As ULong
        If ULong.TryParse(doParse, resultVal) Then
            result = resultVal
            Return True
        End If

        Return False
    End Function
End Class

File: UserInterface/CommandsCommon.cs
using BirthdayBot.Data;
using NodaTime;
using System.Collections.ObjectModel;
using System.Text.RegularExpressions;

namespace BirthdayBot.UserInterface;

/// <summary>
/// Common base class for common constants and variables.
/// </summary>
internal abstract class CommandsCommon {
#if DEBUG
    public const string CommandPrefix = "bt.";
#else
        public const string CommandPrefix = "bb.";
#endif
    public const string BadUserError = ":x: Unable to find user. Specify their `@` mention or their ID.";
    public const string ParameterError = ":x: Invalid usage. Refer to how to use the command and try again.";
    public const string NoParameterError = ":x: This command does not accept any parameters.";
    public const string InternalError = ":x: An unknown error occurred. If it persists, please notify the bot owner.";
    public const string MemberCacheEmptyError = ":warning: Please try the command again.";

    public delegate Task CommandHandler(ShardInstance instance, GuildConfiguration gconf,
                                        string[] param, SocketTextChannel reqChannel, SocketGuildUser reqUser);

    protected static ReadOnlyDictionary<string, string> TzNameMap { get; }
    protected static Regex ChannelMention { get; } = new Regex(@"<#(\d+)>");
    protected static Regex UserMention { get; } = new Regex(@"\!?(\d+)>");

    protected Configuration BotConfig { get; }

    protected CommandsCommon(Configuration db) {
        BotConfig = db;
    }

    static CommandsCommon() {
        var dict = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
        foreach (var name in DateTimeZoneProviders.Tzdb.Ids) dict.Add(name, name);
        TzNameMap = new(dict);
    }

    /// <summary>
    /// On command dispatcher initialization, it will retrieve all available commands through here.
    /// </summary>
    public abstract IEnumerable<(string, CommandHandler)> Commands { get; }

    /// <summary>
    /// Checks given time zone input. Returns a valid string for use with NodaTime.
    /// </summary>
    protected static string ParseTimeZone(string tzinput) {
        if (tzinput.Equals("Asia/Calcutta", StringComparison.OrdinalIgnoreCase)) tzinput = "Asia/Kolkata";
        if (!TzNameMap.TryGetValue(tzinput, out string? tz)) throw new FormatException(":x: Unexpected time zone name."
                     + $" Refer to `{CommandPrefix}help-tzdata` to help determine the correct value.");
        return tz;
    }

    /// <summary>
    /// Given user input where a user-like parameter is expected, attempts to resolve to an ID value.
    /// Input must be a mention or explicit ID. No name resolution is done here.
    /// </summary>
    protected static bool TryGetUserId(string input, out ulong result) {
        string doParse;
        var m = UserMention.Match(input);
        if (m.Success) doParse = m.Groups[1].Value;
        else doParse = input;

        if (ulong.TryParse(doParse, out ulong resultVal)) {
            result = resultVal;
            return true;
        }

        result = default;
        return false;
    }

    /// <summary>
    /// An alternative to <see cref="SocketGuild.HasAllMembers"/> to be called by command handlers needing a full member cache.
    /// Creates a download request if necessary.
    /// </summary>
    /// <returns>
    /// True if the member cache is already filled, false otherwise.
    /// </returns>
    /// <remarks>
    /// Any updates to the member cache aren't accessible until the event handler finishes execution, meaning proactive downloading
    /// is necessary, and is handled by <seealso cref="BackgroundServices.AutoUserDownload"/>. In situations where
    /// this approach fails, this is to be called, and the user must be asked to attempt the command again if this returns false.
    /// </remarks>
    protected static async Task<bool> HasMemberCacheAsync(SocketGuild guild) {
        if (Common.HasMostMembersDownloaded(guild)) return true;
        // Event handling thread hangs if awaited normally or used with Task.Run
        await Task.Factory.StartNew(guild.DownloadUsersAsync).ConfigureAwait(false);
        return false;
    }
}
